api/src/main/java/io/opentelemetry/api/baggage/ImmutableBaggage.java

```diff
@@ -160,38 +97,48 @@ public String toString() {
 
     @Override
     public Baggage.Builder put(String key, String value, EntryMetadata entryMetadata) {
-      entries.put(
-          Objects.requireNonNull(key, "key"),
-          Entry.create(
-              key,
-              Objects.requireNonNull(value, "value"),
-              Objects.requireNonNull(entryMetadata, "entryMetadata")));
+      requireNonNull(key, "key");
+      requireNonNull(value, "value");
+      requireNonNull(entryMetadata, "entryMetadata");
+
+      data.add(key);
+      data.add(Entry.create(key, value, entryMetadata));
       return this;
     }
 
     @Override
     public Baggage.Builder put(String key, String value) {
-      entries.put(
-          Objects.requireNonNull(key, "key"),
-          Entry.create(key, Objects.requireNonNull(value, "value"), EntryMetadata.EMPTY));
-      return this;
+      requireNonNull(key, "key");
+      requireNonNull(value, "value");
+      return put(key, value, EntryMetadata.EMPTY);
     }
 
     @Override
     public Baggage.Builder remove(String key) {
-      entries.remove(Objects.requireNonNull(key, "key"));
-      if (parent != null && parent.getEntryValue(key) != null) {
-        entries.put(key, null);
-      }
+      requireNonNull(key, "key");
+      data.add(key);
+      data.add(null);
       return this;
     }
 
     @Override
-    public ImmutableBaggage build() {
+    public Baggage build() {
       if (parent == null && !noImplicitParent) {
         parent = Baggage.current();
       }
-      return new ImmutableBaggage(entries, parent);
+
+      List<Object> data = this.data;
+      if (parent != null && !parent.isEmpty()) {
+        List<Object> merged = new ArrayList<>(parent.size() * 2 + data.size());
+        parent.forEach(
+            (key, value, metadata) -> {
+              merged.add(key);
+              merged.add(Entry.create(key, value, metadata));
+            });
+        merged.addAll(data);
```

Comment:
I believe the original implementation intentionally was built not to make copies of all the data, but rely on traversing up the stack. I've been told by @bogdandrutu that this was a very large CPU/memory savings at google. Something to consider before we remove that implementation detail.

Reply:
Yeah but I don't know if we can make it efficient. Previously we were actually copying everything when calling `getEntries`. Now with `forEach`, there's still the questionable `seenKeys` stuff. And I can't think of a way to fix the bug with `equals` without similar copying with the parent approach - actually it's exactly what Sleuth is currently doing to work around our equals being broken :P So the up-front copy seems ok to me.

Reply:
By the way, noticed that opencensus API doesn't actually allow end users to access tags (baggage)

https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/tags/TagContext.java#L42

It appears the default implementation does seem to not use parenting and expose the tags directly, presumably requiring a downcast though not sure.

https://github.com/census-instrumentation/opencensus-java/blob/master/impl_core/src/main/java/io/opencensus/implcore/tags/TagMapImpl.java

It seems this latter one is closer to our spec and is what we want to implement here. I think a custom baggage propagator + baggage implementation should be possible still for those that want to prioritize marshalling at the expense of access speed, but I don't think it should be our default.